File: lib/ai.js
// lib/ai.js
// Unified AI router for NandikaPrep. Routes between DeepSeek and Anthropic Claude.

import { loadState, updateState } from "./storage.js";

const DEEPSEEK_ENDPOINT = "https://api.deepseek.com/chat/completions";
const DEEPSEEK_MODEL = "deepseek-chat";

const ANTHROPIC_ENDPOINT = "https://api.anthropic.com/v1/messages";
const ANTHROPIC_MODEL = "claude-sonnet-4-20250514";
const ANTHROPIC_VERSION = "2023-06-01";

const DEEPSEEK_TIMEOUT_MS = 20000;

// Module-level variable tracking the last successful provider
let lastProvider = null;

/**
 * Wraps fetch with a timeout. Returns the response or throws on timeout/error.
 * @param {string} url
 * @param {RequestInit} options
 * @param {number} timeoutMs
 * @returns {Promise<Response>}
 */
async function fetchWithTimeout(url, options, timeoutMs) {
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), timeoutMs);
  try {
    const res = await fetch(url, { ...options, signal: controller.signal });
    return res;
  } finally {
    clearTimeout(timer);
  }
}

/**
 * Calls DeepSeek chat completions API (OpenAI format).
 * @param {string} apiKey
 * @param {string} system
 * @param {string} user
 * @param {boolean} json
 * @returns {Promise<string|null>}
 */
async function callDeepSeek(apiKey, system, user, json) {
  const messages = [];
  if (system) messages.push({ role: "system", content: system });
  messages.push({ role: "user", content: user });

  const body = {
    model: DEEPSEEK_MODEL,
    messages,
  };
  if (json) {
    body.response_format = { type: "json_object" };
  }

  const res = await fetchWithTimeout(
    DEEPSEEK_ENDPOINT,
    {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        Authorization: "Bearer " + apiKey,
      },
      body: JSON.stringify(body),
    },
    DEEPSEEK_TIMEOUT_MS
  );

  if (!res.ok) {
    throw new Error("DeepSeek HTTP " + res.status);
  }

  const data = await res.json();
  return data?.choices?.[0]?.message?.content ?? null;
}

/**
 * Builds Anthropic message content from text, image, and pdf params.
 * @param {string} user
 * @param {string|null} image - base64 data URL
 * @param {string|null} pdf - base64 data URL
 * @returns {Array}
 */
function buildAnthropicContent(user, image, pdf) {
  const content = [];

  if (image) {
    // image is expected as a data URL: "data:<mediaType>;base64,<data>"
    const match = image.match(/^data:([^;]+);base64,(.+)$/);
    if (match) {
      content.push({
        type: "image",
        source: {
          type: "base64",
          media_type: match[1],
          data: match[2],
        },
      });
    }
  }

  if (pdf) {
    const match = pdf.match(/^data:([^;]+);base64,(.+)$/);
    if (match) {
      content.push({
        type: "document",
        source: {
          type: "base64",
          media_type: "application/pdf",
          data: match[2],
        },
      });
    }
  }

  if (user) {
    content.push({ type: "text", text: user });
  }

  return content;
}

/**
 * Calls Anthropic Claude API.
 * @param {string} apiKey
 * @param {string} system
 * @param {string} user
 * @param {boolean} json
 * @param {string|null} image
 * @param {string|null} pdf
 * @returns {Promise<string|null>}
 */
async function callAnthropic(apiKey, system, user, json, image, pdf) {
  const content = buildAnthropicContent(user, image, pdf);

  const body = {
    model: ANTHROPIC_MODEL,
    max_tokens: 2048,
    messages: [{ role: "user", content }],
  };

  if (system) {
    body.system = system;
  }

  const res = await fetch(ANTHROPIC_ENDPOINT, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      "x-api-key": apiKey,
      "anthropic-version": ANTHROPIC_VERSION,
      "anthropic-dangerous-direct-browser-access": "true",
    },
    body: JSON.stringify(body),
  });

  if (!res.ok) {
    throw new Error("Anthropic HTTP " + res.status);
  }

  const data = await res.json();
  const text = data?.content?.[0]?.text ?? null;
  return text;
}

/**
 * callAI is the unified AI router.
 *
 * @param {object} params
 * @param {string} [params.task] - Optional task label for logging
 * @param {string} [params.system] - System prompt
 * @param {string} [params.user] - User message
 * @param {boolean} [params.json] - If true, parse response as JSON
 * @param {string|null} [params.image] - Base64 data URL for vision
 * @param {string|null} [params.pdf] - Base64 data URL for PDF vision
 * @returns {Promise<string|object|null>}
 */
export async function callAI({ task, system = "", user = "", json = false, image = null, pdf = null } = {}) {
  const state = loadState();
  const { deepseek: deepseekKey, anthropic: anthropicKey } = state.apiKeys ?? {};

  // Vision path: Claude only
  if (image || pdf) {
    if (!anthropicKey) {
      return { error: "Vision unavailable, please retry or enter answers manually." };
    }
    try {
      updateState({ providerStats: { ...state.providerStats, visionCalls: (state.providerStats.visionCalls ?? 0) + 1 } });
      const text = await callAnthropic(anthropicKey, system, user, json, image, pdf);
      updateState({
        providerStats: {
          ...(loadState().providerStats),
          anthropicSuccess: (loadState().providerStats.anthropicSuccess ?? 0) + 1,
        },
      });
      lastProvider = "anthropic";
      if (json && text) {
        try { return JSON.parse(text); } catch { return text; }
      }
      return text;
    } catch {
      updateState({
        providerStats: {
          ...(loadState().providerStats),
          anthropicFail: (loadState().providerStats.anthropicFail ?? 0) + 1,
        },
      });
      lastProvider = null;
      return { error: "Vision unavailable, please retry or enter answers manually." };
    }
  }

  // Text path: DeepSeek first, then Claude fallback
  if (deepseekKey) {
    let attempts = 0;
    while (attempts < 2) {
      attempts++;
      try {
        const text = await callDeepSeek(deepseekKey, system, user, json);
        updateState({
          providerStats: {
            ...(loadState().providerStats),
            deepseekSuccess: (loadState().providerStats.deepseekSuccess ?? 0) + 1,
          },
        });
        lastProvider = "deepseek";
        if (json && text) {
          try { return JSON.parse(text); } catch { return text; }
        }
        return text;
      } catch {
        // retry once more, then fall through to Claude
        if (attempts >= 2) {
          updateState({
            providerStats: {
              ...(loadState().providerStats),
              deepseekFail: (loadState().providerStats.deepseekFail ?? 0) + 1,
            },
          });
        }
      }
    }
  }

  // Claude fallback for text
  if (anthropicKey) {
    try {
      const text = await callAnthropic(anthropicKey, system, user, json, null, null);
      updateState({
        providerStats: {
          ...(loadState().providerStats),
          anthropicSuccess: (loadState().providerStats.anthropicSuccess ?? 0) + 1,
        },
      });
      lastProvider = "anthropic";
      if (json && text) {
        try { return JSON.parse(text); } catch { return text; }
      }
      return text;
    } catch {
      updateState({
        providerStats: {
          ...(loadState().providerStats),
          anthropicFail: (loadState().providerStats.anthropicFail ?? 0) + 1,
        },
      });
    }
  }

  // Both failed
  updateState({
    providerStats: {
      ...(loadState().providerStats),
      offlineFallback: (loadState().providerStats.offlineFallback ?? 0) + 1,
    },
  });
  lastProvider = null;
  return null;
}

/**
 * getProviderBadge returns a display string based on the last provider used.
 * @returns {"Reviewed by DeepSeek"|"Reviewed by Claude"|"Reviewed offline"}
 */
export function getProviderBadge() {
  if (lastProvider === "deepseek") return "Reviewed by DeepSeek";
  if (lastProvider === "anthropic") return "Reviewed by Claude";
  return "Reviewed offline";
}

File: lib/ai.py
import base64
import json
import sys

from shared import ANTHROPIC_API_KEY, DEEPSEEK_API_KEY, CLAUDE_MODEL, DEEPSEEK_MODEL


class VisionUnavailable(Exception):
    pass


def _log(msg: str, exc: Exception | None = None) -> None:
    parts = [f"[ai.py] {msg}"]
    if exc:
        parts.append(str(exc))
    print(" | ".join(parts), file=sys.stderr)


def _get_deepseek_client():
    from openai import OpenAI
    return OpenAI(api_key=DEEPSEEK_API_KEY, base_url="https://api.deepseek.com")


def _get_claude_client():
    from anthropic import Anthropic
    return Anthropic(api_key=ANTHROPIC_API_KEY)


def call_ai(
    system: str,
    user: str,
    json_mode: bool = False,
    image: bytes | None = None,
    pdf_pages: list[bytes] | None = None,
    max_tokens: int = 4096,
    db=None,
    user_id: int | None = None,
) -> tuple[str | None, str]:
    """
    Returns (response_text, provider_used).
    Vision path: Claude only.
    Text path: DeepSeek primary, Claude fallback, then (None, 'offline').
    """
    if image or pdf_pages:
        try:
            text = _call_claude_vision(system, user, image, pdf_pages, max_tokens=max_tokens)
            _bump_stats(db, user_id, "claude", success=True, vision=True)
            return text, "claude"
        except Exception as e:
            _bump_stats(db, user_id, "claude", success=False, vision=True)
            raise VisionUnavailable(f"Claude vision failed: {e}") from e

    try:
        text = _call_deepseek(system, user, json_mode, max_tokens=max_tokens)
        _bump_stats(db, user_id, "deepseek", success=True)
        return text, "deepseek"
    except Exception as e:
        _bump_stats(db, user_id, "deepseek", success=False)
        _log("deepseek failed", e)

    try:
        text = _call_claude_text(system, user, json_mode, max_tokens=max_tokens)
        _bump_stats(db, user_id, "claude", success=True)
        return text, "claude"
    except Exception as e:
        _bump_stats(db, user_id, "claude", success=False)
        _log("claude failed", e)

    return None, "offline"


def call_ai_json(
    system: str,
    user: str,
    max_tokens: int = 4096,
    db=None,
    user_id: int | None = None,
) -> tuple[dict | None, str]:
    """Convenience wrapper that parses the JSON response."""
    text, provider = call_ai(system, user, json_mode=True, max_tokens=max_tokens, db=db, user_id=user_id)
    if text is None:
        return None, provider
    try:
        return json.loads(text), provider
    except json.JSONDecodeError:
        _log("JSON parse failed", None)
        return None, "offline"


def _call_deepseek(system: str, user: str, json_mode: bool, max_tokens: int) -> str:
    client = _get_deepseek_client()
    kwargs: dict = {
        "model": DEEPSEEK_MODEL,
        "messages": [
            {"role": "system", "content": system},
            {"role": "user", "content": user},
        ],
        "timeout": 30,
        "max_tokens": max_tokens,
    }
    if json_mode:
        kwargs["response_format"] = {"type": "json_object"}
    resp = client.chat.completions.create(**kwargs)
    return resp.choices[0].message.content or ""


def _call_claude_text(system: str, user: str, json_mode: bool, max_tokens: int) -> str:
    client = _get_claude_client()
    sys_msg = system
    if json_mode:
        sys_msg += "\n\nOutput strictly valid JSON. Output nothing outside the JSON."
    resp = client.messages.create(
        model=CLAUDE_MODEL,
        max_tokens=max_tokens,
        system=sys_msg,
        messages=[{"role": "user", "content": user}],
        timeout=30,
    )
    return resp.content[0].text


def _call_claude_vision(
    system: str,
    user: str,
    image: bytes | None,
    pdf_pages: list[bytes] | None,
    max_tokens: int,
) -> str:
    client = _get_claude_client()
    content_blocks: list[dict] = []
    images = pdf_pages if pdf_pages else ([image] if image else [])
    for img_bytes in images:
        b64 = base64.standard_b64encode(img_bytes).decode("utf-8")
        content_blocks.append({
            "type": "image",
            "source": {"type": "base64", "media_type": "image/jpeg", "data": b64},
        })
    content_blocks.append({"type": "text", "text": user})
    resp = client.messages.create(
        model=CLAUDE_MODEL,
        max_tokens=max_tokens,
        system=system,
        messages=[{"role": "user", "content": content_blocks}],
        timeout=60,
    )
    return resp.content[0].text


def _bump_stats(db, user_id: int | None, provider: str, success: bool, vision: bool = False) -> None:
    if db is None or user_id is None:
        return
    try:
        from lib.db import ProviderStat
        row = db.query(ProviderStat).filter_by(user_id=user_id, provider=provider).first()
        if not row:
            row = ProviderStat(user_id=user_id, provider=provider)
            db.add(row)
        if success:
            row.success_count = (row.success_count or 0) + 1
        else:
            row.fail_count = (row.fail_count or 0) + 1
        if vision:
            row.vision_call_count = (row.vision_call_count or 0) + 1
        db.commit()
    except Exception as e:
        _log("_bump_stats failed", e)
